feat(probabilistic_occupancy_grid_map): add diagnostics warning when latency is too long

ogrex opened the pull request:
## Description
Launcher PR
https://github.com/autowarefoundation/autoware_launch/pull/1387

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?
Tested by running logging simulator

OK:
![image](https://github.com/user-attachments/assets/be1137a7-3295-408b-9a81-778a33d24545)



Warnings:
![image](https://github.com/user-attachments/assets/7162c57c-e6c0-4bf0-bd03-f619cbfbaf95)
![image](https://github.com/user-attachments/assets/a5ae17f4-11dd-4cda-a448-ea3a85cabc0b)


## Notes for reviewers

None.

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

technolojin reviewed (commented):
similar to https://github.com/autowarefoundation/autoware.universe/pull/10257/files#, it would be better to set tolerance duration of `is_delay_within_range=false`

technolojin reviewed (commented):
similar to https://github.com/autowarefoundation/autoware.universe/pull/10257/files, add a tolerance duration of `is_delay_within_range=false` would be better

knzo25 commented on perception/autoware_probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml:
```diff
@@ -35,3 +35,4 @@
 
     # debug parameters
     publish_processing_time_detail: false
+    max_output_delay_ms: 50.0
```
Please update the schema as well

knzo25 commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -145,6 +145,10 @@ PointcloudBasedOccupancyGridMapNode::PointcloudBasedOccupancyGridMapNode(
     }
   }
+
+  max_output_delay_ms_ = this->declare_parameter<double>("max_output_delay_ms", 50.0);
```
We are trying to avoid default values for the parameters

knzo25 commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -270,6 +274,22 @@ void PointcloudBasedOccupancyGridMapNode::onPointcloudWithObstacleAndRaw()
+    diagnostics_interface_ptr_->add_key_value(
+      "is_output_delay_within_range", is_delay_within_range);
+
+    std::stringstream message;
```
Although it compiles, pre-commit should have warned about including the sstream header as well

knzo25 commented:
@ogrex 
In this repository, please remember to add the tag `run:build-and-test-differential` when you open your PRs. It will trigger some of the CI/CD jobs that are required before merging the PR

knzo25 reviewed (changes requested):
Left some comments, so please check them later

knzo25 commented:
@ogrex 
FYI, it is customary to paste in the review comment the commit hash which implemented the required change. That way the review can focus on the new changes rather than reviewing the whole PR again

knzo25 commented:
@technolojin @ogrex 
This PR uses latency instead of processing time. Is that on purpose?
Latency at this node should be at around 180-200ms no (some of the preprocessing involves: 100ms of lidar, 20 ms of waiting, 40ms of preprocessing, 10ms of segmentation, 10ms of subsample, 5ms of OGM)?

The implementation also differs from https://github.com/autowarefoundation/autoware_universe/pull/10284
Is that Ok?

ogrex commented:
@knzo25  
3f263b6aac971797de20adc024293829c024fa23 
Updated the commit with 
1. Introduced a duration tolerance check. Since the implementation code has been refactored, please review it again based on this. 
2. Changed target latency to processing_time_ms and make set values align with it , 
 in case pipeline latency is necessary I will change it back

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml:
```diff
@@ -35,3 +35,5 @@
 
     # debug parameters
     publish_processing_time_detail: false
+    max_output_delay_ms: 50.0 # [ms]
```
IMO, `output_delay` is ambiguous. 
- delay from measurement to result?
- delay of processing?

```suggestion
# diagnostics parameters
    processing_time_tolerance_ms: 50.0 # [ms]
    processing_time_consecutive_excess_tolerance_ms: 1000.0 # [ms]
```

ogrex replied:
just confirm , target is processing_time_ms,  correct?

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/schema/pointcloud_based_occupancy_grid_map.schema.json:
```diff
@@ -125,6 +125,16 @@
+            "max_acceptable_consecutive_delay_ms": {
+              "type": "number",
+              "description": "Maximum acceptable consecutive delay warning in milliseconds.",
+              "default": 50.0
```
default value is not aligned

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -148,6 +149,12 @@ PointcloudBasedOccupancyGridMapNode::PointcloudBasedOccupancyGridMapNode(
   }
+
+  max_output_delay_ms_ = this->declare_parameter<double>("max_output_delay_ms");
+  max_acceptable_consecutive_delay_ms_ =
```
rename it to follow the parameter

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -170,6 +177,52 @@ void PointcloudBasedOccupancyGridMapNode::rawPointcloudCallback(
+void PointcloudBasedOccupancyGridMapNode::checkLatency(double current_latency_ms)
+{
+  static rclcpp::Time last_normal_time = this->get_clock()->now();
+  const bool is_delay_within_range = (current_latency_ms <= max_output_delay_ms_);
```
also rename it as `processing time`

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -170,6 +177,52 @@ void PointcloudBasedOccupancyGridMapNode::rawPointcloudCallback(
+  }
+
+  diagnostics_interface_ptr_->clear();
+  diagnostics_interface_ptr_->add_key_value("latency(ms)", current_latency_ms);
```
processing time

technolojin commented on perception/autoware_probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/pointcloud_based_occupancy_grid_map_node.cpp:
```diff
@@ -170,6 +177,52 @@ void PointcloudBasedOccupancyGridMapNode::rawPointcloudCallback(
+  diagnostics_interface_ptr_->clear();
+  diagnostics_interface_ptr_->add_key_value("latency(ms)", current_latency_ms);
+  diagnostics_interface_ptr_->add_key_value("is_latency_within_threshold", is_delay_within_range);
+  diagnostics_interface_ptr_->add_key_value("abnormal_duration(ms)", abnormal_duration_ms);
```
processing time - consecutive excess duration

technolojin reviewed (changes requested):
I would like to request parameters and variables renaming.

reference: https://github.com/autowarefoundation/autoware_universe/pull/10219/files

technolojin reviewed (approved):
LGTM

(In terms of the diagnostics implementation)

Dear code owners, I would like to ask whether the tolerance values are appropriate.

technolojin commented:
@YoshiRi 
Do you think processing_time_tolerance_ms: 50.0 is acceptable? I would like to ask to your experience.

knzo25 reviewed (approved):
LGTM.
If several nodes do the same, It would be better to make the same logic available to all nodes imo

technolojin commented:
average processing time of this module is 10ms. 50ms of threshold is large enough.

technolojin commented:
Missing parameter for multi-lidar mode

https://github.com/autowarefoundation/autoware_universe/pull/10578
https://github.com/autowarefoundation/autoware_launch/pull/1426
